[SaferCPP] Eliminate raw pointer hash table for Thread

geoffreygaren opened the pull request:
#### 26ebb2b1983fed45ad687a87c8d0863f47f6537d
<pre>
[SaferCPP] Eliminate raw pointer hash table for Thread
<a href="https://bugs.webkit.org/show_bug.cgi?id=302414">https://bugs.webkit.org/show_bug.cgi?id=302414</a>
<a href="https://rdar.apple.com/164575958">rdar://164575958</a>

Reviewed by Brady Eidson and Chris Dumez.

Adopted ThreadSafeWeakHashSet instead. It&apos;s pretty neat!

Now we end up locking less than we used to, and that revealed a race
condition between thread startup and shutdown in workers/bomb.html.

Fixed the race condition by ensuring that all Thread initialization happens
before launching the new thread or while holding the NewThreadContext lock
(either on the caller thread or on the callee thread, depending on build config).

Canonical link: <a href="https://commits.webkit.org/303075@main">https://commits.webkit.org/303075@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d572f5d1888a8c64e1e403334f5f96fdc1cd3264

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/131206 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/3536 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/42170 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/138649 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/82916 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/40845ec5-2445-4450-81c3-cca094491977) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/133076 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/3534 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/3378 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99965 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/67715 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/b7c31ca6-ac53-41c7-a2ee-80eba3ecf211) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134152 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/2532 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/117453 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/80665 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/cdcdf00c-155c-4f9b-ba44-cf5c96f75673) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/2448 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/152 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/81898 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/123228 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/111042 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/35573 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/141148 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/129660 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/3281 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/36098 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/108486 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/3326 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/2933 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/108429 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2464 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/113785 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/56361 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20414 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/3343 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/32209 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/162677 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/3165 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/66750 "Built successfully") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/40876 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/3365 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/3273 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

beidson commented on Tools/TestWebKitAPI/Tests/WTF/RunLoop.cpp:
```diff
@@ -274,8 +274,7 @@ TEST(WTF_RunLoop, Create)
-        Locker threadsLock { Thread::allThreadsLock() };
-        EXPECT_TRUE(Thread::allThreads().contains(runLoopThread));
+        EXPECT_TRUE(Thread::allThreads().contains(*runLoopThread));
     }
```
Can remove the scope block now.

beidson commented on Tools/TestWebKitAPI/Tests/WTF/RunLoop.cpp:
```diff
@@ -286,8 +285,7 @@ TEST(WTF_RunLoop, Create)
-        Locker threadsLock { Thread::allThreadsLock() };
-        EXPECT_FALSE(Thread::allThreads().contains(runLoopThread));
+        EXPECT_FALSE(Thread::allThreads().contains(*runLoopThread));
     }
```
Can remove the scope block now

cdumez commented on Source/WTF/wtf/Threading.cpp:
```diff
@@ -271,44 +271,32 @@ void Thread::entryPoint(NewThreadContext* newThreadContext)
-    // and make the context ThreadSafeRefCounted.
-    context->ref();
+
+    Ref<Thread> thread = adoptRef(*new Thread(schedulingPolicy));
```
nit: `Ref thread = ...`

cdumez commented on Source/WTF/wtf/Threading.cpp:
```diff
@@ -271,44 +271,32 @@ void Thread::entryPoint(NewThreadContext* newThreadContext)
+
+    Ref<Thread> thread = adoptRef(*new Thread(schedulingPolicy));
+
+    Ref<NewThreadContext> context = adoptRef(*new NewThreadContext { name, WTFMove(entryPoint), thread.get() });
```
nit : `Ref context`

cdumez commented on Source/WTF/wtf/Threading.h:
```diff
@@ -314,12 +313,15 @@ class WTF_CAPABILITY("is current") Thread : public ThreadSafeRefCounted<Thread>,
 
 protected:
-    Thread() = default;
+    Thread(SchedulingPolicy schedulingPolicy)
```
nit: `explicit`?

cdumez commented on Source/WTF/wtf/posix/ThreadingPOSIX.cpp:
```diff
@@ -437,7 +437,7 @@ Thread& Thread::initializeCurrentTLS()
     // Not a WTF-created thread, Thread is not established yet.
     WTF::initialize();
-    Ref<Thread> thread = adoptRef(*new Thread());
+    Ref<Thread> thread = adoptRef(*new Thread(SchedulingPolicy::Other));
```
`Ref thread`

cdumez commented on Source/WTF/wtf/win/ThreadingWin.cpp:
```diff
@@ -231,7 +231,7 @@ Thread& Thread::initializeCurrentTLS()
     // Not a WTF-created thread, ThreadIdentifier is not established yet.
     WTF::initialize();
-    Ref<Thread> thread = adoptRef(*new Thread());
+    Ref<Thread> thread = adoptRef(*new Thread(SchedulingPolicy::Other));
```
`Ref thread`

geoffreygaren committed:
[SaferCPP] Eliminate raw pointer hash table for Thread
https://bugs.webkit.org/show_bug.cgi?id=302414
rdar://164575958

Reviewed by Brady Eidson and Chris Dumez.

Adopted ThreadSafeWeakHashSet instead. It's pretty neat!

Now we end up locking less than we used to, and that revealed a race
condition between thread startup and shutdown in workers/bomb.html.

Fixed the race condition by ensuring that all Thread initialization happens
before launching the new thread or while holding the NewThreadContext lock
(either on the caller thread or on the callee thread, depending on build config).

Canonical link: https://commits.webkit.org/303075@main
